pkg/authentication/middlewares.go

```diff
@@ -38,29 +48,82 @@ func (m *Middleware) isAllowListed(r *http.Request) bool {
 	return false
 }
 
-func (m *Middleware) OAuth2Authentication(next http.Handler) http.Handler {
+func (m *Middleware) OAuth2AuthenticationChain() []func(http.Handler) http.Handler {
+	return []func(http.Handler) http.Handler{
+		m.OAuth2BearerAuthentication,
+		m.OAuth2CookieAuthentication,
+	}
+}
+
+func (m *Middleware) OAuth2BearerAuthentication(next http.Handler) http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2Authentication")
+		ctx, span := m.tracer.Start(r.Context(), "authentication.Middleware.OAuth2BearerAuthentication")
 		defer span.End()
-		r = r.WithContext(ctx)
 
 		if m.isAllowListed(r) {
-			next.ServeHTTP(w, r.WithContext(ctx))
+			next.ServeHTTP(w, r)
 			return
 		}
 
-		rawAccessToken, err := getBearerToken(r.Header)
-		if err != nil {
-			m.unauthorizedResponse(w, err)
-			return
-		}
+		var (
+			principal *Principal
+			err       error
+		)
 
 		// add the Otel HTTP Client
 		r.WithContext(OtelHTTPClientContext(r.Context()))
 
-		principal, err := m.oauth2.Verifier().VerifyAccessToken(r.Context(), rawAccessToken)
+		if rawAccessToken, found := m.getBearerToken(r.Header); found {
+			principal, err = m.oauth2.Verifier().VerifyAccessToken(r.Context(), rawAccessToken)
```

Comment:
This has been already taken care of in one of my previous PR (following your suggestion 😎) - [link to code](https://github.com/canonical/identity-platform-admin-ui/blob/main/pkg/authentication/tokens.go#L91-L106)
The validation is that the audience value in the userinfo response for the access token must match exactly the client id we use for minting the tokens.

The userinfo has the audience `aud` field in the response (since the client is configured to allow the audience value we want, and when the authentication request had the additional `audience` parameter specified, both of these are strong requirements for our solution)

The id token is always a JWT and the audience check is performed for it also, so we're good.

Reply:
I opened an issue for this improvement - https://github.com/canonical/identity-platform-admin-ui/issues/342